### DISCUSSION TITLE
Better interface modeling for VMs along with physical interface mapping to hypervisor interfaces

### DISCUSSION THREAD

**Discussion by eronlloyd** [netbox-community/netbox#12621] [Ideas]
Perhaps I can't find the conversation discussing this, but it would be nice to be able to specify VM interface speeds. This seems to be one of the only things (maybe also duplex mode), that is still relevant for virtual NOS instances.

This is how the docs discuss the VMInterface model:

> [Virtual machine](https://192.168.1.7/static/docs/models/virtualization/virtualmachine/) [interfaces](https://192.168.1.7/static/docs/models/dcim/interface/) behave similarly to device interfaces: They can be assigned to VRFs, may have IP addresses, VLANs, and services attached to them, and so on. However, given their virtual nature, they lack properties pertaining to physical attributes. For example, VM interfaces do not have a physical type and cannot have cables attached to them.

I also have been thinking it would be nice to be able to model hypervisor virtual switches/bridges and map them to the VMInterface instances to track that at both the physical and virtual levels. I have to think this through a bit more, but wanted to plant a seed on it to see if anyone else was thinking the same thing.

**Comment by candlerb** [comment]
Which virtualization environment are you using that allows you to specify "interface speed" on a virtual interface?

Is it actually some sort of software rate limiting (so you could set, say 50Mbps or 300Mbps)? In that case, mapping to fixed rates like 1G and 10G wouldn't work. I'd suggest you set a custom field on those interfaces instead.

Duplex mode absolutely doesn't apply to VM interfaces. It's a layer 2 attribute of a physical interface: whether you can send a packet onto the transmission medium concurrently with receiving a packet.

**Reply by eronlloyd** [reply]
My main use case is trying to model VM interface mapping to the underlying hypervisor vSwitch or Linux bridge, so for instance knowing which Juniper JunosOS VMs are utilizing the SFP+ interfaces vs 1000BaseT at the physical layer.

Say I have a hypervisor with many VMs, including a vSRX acting as a LAN firewall and a Linux VM acting as a speed test server; I want to ensure the speed test server maintains enough bandwidth to perform its throughput-sensitive service, whereas the firewall and other VMs don't care. I'd like to track hypervisor interface allocations to the VM interface overlays to monitor over-subscription of the physical ports. I would imagine this increasingly becomes a concern with more NFV designs.

I took a closer look at how a Juniper vSRX is representing its interfaces on a hypervisor with both 1000BaseT and 10GE SFP+ ports, and interestingly all vSRX interfaces are reporting 10Gbps even though the underlying Linux bridge only has a single 1000BaseT interface.

Seems like I should dig into bridging further before revisiting this. Thanks for helping me refine my goal here, @candlerb.
